experimental/harfbuzz/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+        .get32(unicode) as usize] as hb_unicode_general_category_t
+}
+
+unsafe extern "C" fn icu4x_hb_unicode_general_category_destroy(user_data: *mut c_void) {
+    let _ = Box::from_raw(user_data as *mut CodePointMapData<GeneralCategory>);
+}
+
+unsafe extern "C" fn icu4x_hb_unicode_compose(
+    _ufuncs: *mut hb_unicode_funcs_t,
+    a: hb_codepoint_t,
+    b: hb_codepoint_t,
+    ab: *mut hb_codepoint_t,
+    user_data: *mut c_void,
+) -> hb_bool_t {
+    // Could we trust HarfBuzz to pass valid scalar values?
+    // `hb_buffer_add_codepoints` seems to allow the introduction
+    // of non-scalar-value items to the HarfBuzz buffer. Should we
+    // treat that as conceptually `unsafe` so that if the app has
+    // introduced values that are not valid scalar values, we'd
+    // be OK with UB? OTOH, HarfBuzz docs say "code points", so
+    // following docs to the letter could still result in
+    // surrogate code points even if they don't make sense.
+    let first = if let Some(first) = core::char::from_u32(a) {
+        first
+    } else {
+        return false as hb_bool_t;
+    };
+    let second = if let Some(second) = core::char::from_u32(b) {
+        second
+    } else {
+        return false as hb_bool_t;
+    };
+    if let Some(c) = (*(user_data as *mut CanonicalComposition)).compose(first, second) {
+        core::ptr::write(ab, c as hb_codepoint_t);
+        true as hb_bool_t
+    } else {
+        false as hb_bool_t
+    }
+}
+
+unsafe extern "C" fn icu4x_hb_unicode_compose_destroy(user_data: *mut c_void) {
+    let _ = Box::from_raw(user_data as *mut CanonicalComposition);
+}
+
+unsafe extern "C" fn icu4x_hb_unicode_decompose(
+    _ufuncs: *mut hb_unicode_funcs_t,
+    ab: hb_codepoint_t,
+    a: *mut hb_codepoint_t,
+    b: *mut hb_codepoint_t,
+    user_data: *mut c_void,
+) -> hb_bool_t {
+    // Could we trust HarfBuzz to pass a valid scalar value?
+    // `hb_buffer_add_codepoints` seems to allow the introduction
+    // of non-scalar-value items to the HarfBuzz buffer. Should we
+    // treat that as conceptually `unsafe` so that if the app has
+    // introduced values that are not valid scalar values, we'd
+    // be OK with UB? OTOH, HarfBuzz docs say "code points", so
+    // following docs to the letter could still result in
+    // surrogate code points even if they don't make sense.
+    let composed = if let Some(composed) = core::char::from_u32(ab) {
+        composed
+    } else {
+        return false as hb_bool_t;
+    };
+    match (*(user_data as *mut CanonicalDecomposition)).decompose(composed) {
+        Decomposed::Default => false as hb_bool_t,
+        Decomposed::Expansion(first, second) => {
+            core::ptr::write(a, first as hb_codepoint_t);
+            core::ptr::write(b, second as hb_codepoint_t);
+            true as hb_bool_t
+        }
+        Decomposed::Singleton(single) => {
+            core::ptr::write(a, single as hb_codepoint_t);
+            core::ptr::write(b, 0);
```

Comment:
Question: It seems that HarfBuzz does not clearly define what to do with a singleton decomposition; is this the correct behavior, and if so, where did you find it?

Reply:
This is consistent with how Firefox implements this callback and how HarfBuzz uses the callback. See https://searchfox.org/mozilla-central/rev/893a8f062ec6144c84403fbfb0a57234418b89cf/gfx/harfbuzz/src/hb-ot-shape-normalize.cc#156-158 for usage within HarfBuzz.